re-work access handling

jlevon opened the pull request:
Various cleanups and fixes to handling of region accesses, including:

- there should be no reason for us to split accesses into 1/2/4/8 byte accesses:
  in general, the client will have already be doing that, and if not, there's no
  particular reason we should be the ones to split up such larger accesses.

- use a callback for PCI config space reads and writes if one is provided (needs
  more work for capabilities)

Signed-off-by: John Levon <john.levon@nutanix.com>

tmakatos commented on lib/pci.c:
```diff
@@ -265,77 +263,99 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
+    int rc;
+
+    if (offset < PCI_STD_HEADER_SIZEOF) {
+        rc = pci_hdr_access(vfu_ctx, buf, &count, &offset, is_write);
```
Won't this have already been done by `_vfu_access`?

tmakatos reviewed (changes requested):
I've only revied the last commit (re-work access handling) as my understanding is that the othes commits have been reviewed in other PRs.

Regarding no splitting accesses, I don't see the relevant patch. In any case, AFAIK the PCI spec specifies that access must be 1/2/4/8 bytes and aligned. Device implementations might expect this, so if we don' guarantee this then we might end up with device implementations manually doing it. We can't guarantee that a client will do the right thing and this is not specified in the spec. IMO we should keep the current behavior and we can add a switch to disable it if the user prefers to handle larger accesses.

jlevon commented on include/libvfio-user.h:
```diff
@@ -218,42 +218,51 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
+ *  - the @size argument is ignored: the region size is always the size defined
+ *    by the relevant PCI specification
+ *  - all accesses to the standard PCI header are handled by the library
+ *  - all accesses to known PCI capabilities are handled by the library
```
agreed, but I'll do this in the follow on cap add change where we enumerate them

jlevon commented on include/libvfio-user.h:
```diff
@@ -218,42 +218,51 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
+ *
+ *  - the @size argument is ignored: the region size is always the size defined
+ *    by the relevant PCI specification
+ *  - all accesses to the standard PCI header are handled by the library
```
done

jlevon commented on lib/pci.c:
```diff
@@ -216,21 +216,18 @@ handle_erom_write(vfu_ctx_t *ctx, vfu_pci_config_space_t *pci,
-pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
-                  const char *buf, size_t count)
+static int
+pci_hdr_write(vfu_ctx_t *vfu_ctx, vfu_pci_config_space_t *space,
```
done

jlevon commented on lib/pci.c:
```diff
@@ -265,77 +263,99 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
- * @rw: the command
- * @write: whether this is a PCI header write
- * @count: output parameter that receives the number of bytes read/written
- */
```
It's an internal function, and pretty simple, this didn't seem to have any value. We don't do this for all the others?

tmakatos replied:
I tend to add some basic documentation for internal function that aren't dead simple, especially ones where an argument can be misunderstood. I generally try to document even the simplest of functions but sometimes I get lazy :D

jlevon replied:
I'll do so in the top patch - it's changed again there

jlevon commented on lib/pci.c:
```diff
@@ -265,77 +263,99 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
+    size_t count;
     int err = 0;
 
-    assert(vfu_ctx != NULL);
```
I don't personally see much point in NULL checks like this but sure.

jlevon commented on lib/pci.c:
```diff
@@ -265,77 +263,99 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
-    assert(pos != NULL);
-    assert(buf != NULL);
-
-    _pos = *pos - region_to_offset(VFU_PCI_DEV_CFG_REGION_IDX);
```
See the libvfio-user.c changes, this is no longer relevant - all offsets are from the start of the region now

jlevon commented:
I'm still not convinced we need the 1/2/4/8 splitting code (removed by the libvfio-user.c changes). The client should be automatically doing this for any typical access, and other cases (such as non-PCI devices) it seems weird to enforce.

Can we revisit this after this patchset (I have a bunch more access changes that involve splitting)

tmakatos commented:
> I'm still not convinced we need the 1/2/4/8 splitting code (removed by the libvfio-user.c changes). The client should be automatically doing this for any typical access, and other cases (such as non-PCI devices) it seems weird to enforce.
> 
> Can we revisit this after this patchset (I have a bunch more access changes that involve splitting)

Sure that's faily trivial to re-add.

tmakatos commented on lib/libvfio-user.c:
```diff
@@ -188,17 +189,163 @@ dump_buffer(const char *prefix, const char *buf, uint32_t count)
 
+static ssize_t
+region_access(vfu_ctx_t *vfu_ctx, size_t region_index,
+              bool is_write, char *buf, uint64_t offset, size_t count)
```
Can we swap offset and count to make it consistent with the rest of the code? (Which is think matches `pwrite(2)`.)

jlevon replied:
good spot, done

tmakatos commented on lib/libvfio-user.c:
```diff
@@ -188,17 +189,163 @@ dump_buffer(const char *prefix, const char *buf, uint32_t count)
+}
+
+static bool
+valid_region_access(vfu_ctx_t *vfu_ctx, size_t size, uint16_t cmd,
```
I think `validate_region_access` or `is_valid_region_access` would be clearer.

jlevon replied:
done

tmakatos commented on lib/libvfio-user.c:
```diff
@@ -188,17 +189,163 @@ dump_buffer(const char *prefix, const char *buf, uint32_t count)
+{
+    size_t index;
+
+    assert(ra != NULL);
```
We should validate `vfu_ctx` as well (not done in the original code).

jlevon replied:
done

tmakatos commented on lib/libvfio-user.c:
```diff
@@ -188,17 +189,163 @@ dump_buffer(const char *prefix, const char *buf, uint32_t count)
+    }
+
+    if (ra->count == 0) {
+        vfu_log(vfu_ctx, LOG_ERR, "bad region access count of 0");
```
Although it doesn't make sense, it's not prohibited by the spec.

jlevon replied:
done

tmakatos reviewed (commented):
Minor suggestions that can be omitted. This is a big commit so having another pair of eyes (@swapnili's :P) to look at this would be a good idea.

jlevon commented on include/libvfio-user.h:
```diff
@@ -241,6 +223,34 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
  * demonstrate in a sample.
  *
+ * Areas that are accessed via such a mapping by definition do not invoke any
+ * given callback.  However, any access via VFIO_USER_REGION_READ/WRITE will do
```
tried to rewrite this for clarity

jlevon commented on include/libvfio-user.h:
```diff
@@ -241,6 +223,34 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
+ * @region_idx: region index
+ * @size: size of the region
+ * @region_access: callback function to access region
+ * @flags: region flags (VFU_REGION_FLAG_)
```
Yes, but I'm not sure what you mean in context - you'd like me to make changes so this is true?

tmakatos replied:
I think for each of these arguments we must say "ignored if the region is not mappable", rather than e.g. requesting that they have specific values (like we do for `nr_mmap_areas`).

jlevon replied:
actually, I think we should be stricter: if !mappable, they must all be zero/-1. Will catch out silly mistakes easier.

jlevon commented on lib/migration.c:
```diff
@@ -380,14 +381,12 @@ handle_data_size(vfu_ctx_t *vfu_ctx, struct migration *migr,
+    struct migration *migr = vfu_ctx->migration;
     int ret;
 
-    assert(migr != NULL);
```
done

jlevon commented on lib/migration.c:
```diff
@@ -430,18 +429,17 @@ handle_region_access_registers(vfu_ctx_t *vfu_ctx, struct migration *migr,
+                        loff_t pos, bool is_write)
 {
-    int ret = -EINVAL;
+    struct migration *migr = vfu_ctx->migration;
```
done

jlevon commented on lib/pci.c:
```diff
@@ -265,84 +263,109 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
 {
-    int ret;
+    vfu_region_access_cb_t *cb =
+        vfu_ctx->reg_info[VFU_PCI_DEV_CFG_REGION_IDX].fn;
```
done

jlevon commented on lib/pci.c:
```diff
@@ -265,84 +263,109 @@ pci_hdr_write(vfu_ctx_t *vfu_ctx, uint16_t offset,
+ *
+ * Outside of those areas, if a callback is specified for the region, we'll use
+ * that; otherwise, writes are not allowed, and reads are satisfied with
+ * memcpy().
```
done

jlevon commented:
I went a little bit further with the region setup mmap thing - I removed the MMAP flag - it didn't seem to have any value, since the 'fd' parameter defines whether the region is mappable. Also fixed up region info such that we weren't silently relying on matching the VFIO region flag values.

PS I'll wait for Swapnil to take a look

jlevon commented:
@tmakatos sorry, noticed another little fix
